### PR TITLE
DM-9937: Add noexcept specifiers to applicable methods in afw

### PR DESCRIPTION
**Pull request by kfindeisen**
This PR adds `noexcept` specifiers to most functions and methods in `lsst::geom`. I have omitted these specifiers in a couple of key cases:
- any method that directly or indirectly uses dynamic memory allocation, or where it might be reasonable to want that in the future. This includes methods that return `vector` or `string` by value.
- any method that could in principle take an invalid input (e.g., anything that takes an index) has been left potentially throwing
- many of the methods in the `Coordinate` hierarchy, such as `Point::getSquaredNorm`, are in principle `noexcept` but currently have a potentially throwing implementation. I've left these methods unmarked to avoid leaking implementation details.

The exception specs for classes that use Eigen are based on its policy that Eigen classes only throw `bad_alloc`, and handle all other failure modes using `assert`. I don't know if we want to relax this to allow for non-Eigen implementations in the future.

Note that we could simplify the exception specs for `CoordinateBase` quite a bit by using a `static_assert` to force it to use statically allocated Eigen vectors. However, there are plausible coordinate types (e.g., `std::complex`) that are not nothrow-copyable, so most `CoordinateBase` and subclass methods would still be conditionally `noexcept` depending on `T`.

### CODE REVIEW

File: include/lsst/geom/Box.h [hunk 1]
```diff
@@ -118,28 +122,28 @@ class Box2I {
      */
```
Comment (r-owen) [inline include/lsst/geom/Box.h]: Please document the exceptions that this could raise.

File: include/lsst/geom/Box.h [hunk 2]
```diff
@@ -163,34 +167,34 @@ class Box2I {
...
     //@{
-    Extent2I const getDimensions() const { return _dimensions; }
-    int getWidth() const { return _dimensions.getX(); }
-    int getHeight() const { return _dimensions.getY(); }
-    int getArea() const { return getWidth() * getHeight(); }
+    Extent2I const getDimensions() const noexcept { return _dimensions; }
+    int getWidth() const noexcept { return _dimensions.getX(); }
+    int getHeight() const noexcept { return _dimensions.getY(); }
+    int getArea() const noexcept { return getWidth() * getHeight(); }
```
Comment (r-owen) [inline include/lsst/geom/Box.h]: What happens if the area overflows an `int`? I suspect that should raise an exception (even if it does not), so I suggest not adding `noexcept` here.

File: include/lsst/geom/Box.h [hunk 3]
```diff
@@ -163,34 +167,34 @@ class Box2I {
...
     //@}
 
     /// Return slices to extract the box's region from an ndarray::Array.
     ndarray::View<boost::fusion::vector2<ndarray::index::Range, ndarray::index::Range> > getSlices() const;
```
Comment (r-owen) [inline include/lsst/geom/Box.h]: What exception can this raise? It would help to document this as well as you can (even if you don't have details)

File: include/lsst/geom/Box.h [hunk 4]
```diff
@@ -224,21 +228,21 @@ class Box2I {
     void include(Box2I const& other);
 
     /// Shrink this to ensure that other.contains(*this).
-    void clip(Box2I const& other);
+    void clip(Box2I const& other) noexcept;
```
Comment (r-owen) [inline include/lsst/geom/Box.h]: Please consider documenting the behavior in special cases, e.g. if there is no overlap in either dimension then the BBox becomes empty.

(I realize this is pre-existing, but it would be nice to know. And by adding "noexcept" you beg the question of what happens in these special cases)

File: include/lsst/geom/CoordinateBase.h [hunk 5]
```diff
@@ -50,12 +51,26 @@ class CoordinateBase {
...
-    CoordinateBase(CoordinateBase const&) = default;
-    CoordinateBase(CoordinateBase&&) = default;
-    CoordinateBase& operator=(CoordinateBase const&) = default;
-    CoordinateBase& operator=(CoordinateBase&&) = default;
-    ~CoordinateBase() = default;
+    static bool constexpr IS_ELEMENT_NOTHROW_COPYABLE = std::is_nothrow_copy_constructible<T>::value;
+    static bool constexpr IS_VECTOR_NOTHROW_CONSTRUCTIBLE =
+            N != Eigen::Dynamic && IS_ELEMENT_NOTHROW_COPYABLE;
+    static bool constexpr IS_VECTOR_NOTHROW_COPYABLE = N != Eigen::Dynamic && IS_ELEMENT_NOTHROW_COPYABLE;
+    static bool constexpr IS_VECTOR_NOTHROW_ASSIGNABLE = std::is_nothrow_copy_assignable<T>::value;
```
Comment (r-owen) [inline include/lsst/geom/CoordinateBase.h]: I assume this is not required for int or double, the types we use now. What sort of types did you have in mind? Please consider saying something about this in a code comment or the documentation block.

File: include/lsst/geom/CoordinateExpr.h [hunk 6]
```diff
@@ -70,23 +70,23 @@ class CoordinateExpr : public CoordinateBase<CoordinateExpr<N>, bool, N> {
      *  These operators do not provide interoperability with scalars.
      */
     //@{
-    CoordinateExpr and_(CoordinateExpr const& rhs) const;
-    CoordinateExpr or_(CoordinateExpr const& rhs) const;
-    CoordinateExpr not_() const;
+    CoordinateExpr and_(CoordinateExpr const& rhs) const noexcept(noexcept(CoordinateExpr(false)));
+    CoordinateExpr or_(CoordinateExpr const& rhs) const noexcept(noexcept(CoordinateExpr(false)));
```
Comment (r-owen) [inline include/lsst/geom/CoordinateExpr.h]: Why `noexcept(noexcept(...` instead of just `noexcept(...`? If that is a mistake then please look for it globally.

File: include/lsst/geom/Extent.h [hunk 7]
```diff
@@ -154,7 +154,7 @@ class ExtentBase : public CoordinateBase<Extent<T, N>, T, N> {
     //@}
```
Comment (r-owen) [inline include/lsst/geom/Extent.h]: Surely you can mark `operator==` and `operator!=` as `noextent`?

I assume the math operators are out because they return a new `Extent`.

**Comment by TallJimbo** [comment]
FWIW, I would be happy to `static_assert` that `CoordinateBase`'s template parameters are scalar numeric types, and rule out `std::complex` or anything else that lacks nothrow construction.

And I'd definitely be in favor of adding a `static_assert` requiring statically-allocated Eigen objects.  I actually thought that was already guaranteed by the class design.

**Comment by kfindeisen** [comment]
> I would be happy to static_assert that CoordinateBase's template parameters are scalar numeric types, and rule out std::complex or anything else that lacks nothrow construction.

I think restricting the coordinates to fundamental types (let's not say "scalar", since the C++ standard uses a different meaning from the mathematical one and that could cause confusion) would require an RFC, even though AFAIK no existing code would be broken. It would be a significant restriction on what one can do with these templates in the future.

> there are a lot of new unconditional noexcept qualifiers in afw on this ticket that should be conditional on the Point and Extent constructors being noexcept as well.

The conditions are satisfied: Point2D, Point2I, Point3D, and Point3I are `noexcept` constructible, copyable, etc.. It's only when we have unknown template arguments that we need to worry about the conditions.

**Comment by TallJimbo** [comment]
Oh, of course, the templates are conditionally `noexcept`, but the instantiations used in afw are unconditionally `noexcept`.  Sorry for the noise.

File: include/lsst/geom/Box.h (hunk 3 above)
Comment (kfindeisen) [reply include/lsst/geom/Box.h]: I honestly have no idea. Since Boost rarely documents its exceptions, I decided it wasn't worth investigating.

File: include/lsst/geom/CoordinateExpr.h (hunk 6 above)
Comment (kfindeisen) [reply include/lsst/geom/CoordinateExpr.h]: That is the actual (and, I agree, very confusing) syntax. It's why I preferred to use variables where practical.

`noexcept()` [used as a function/method keyword](https://en.cppreference.com/w/cpp/language/noexcept_spec) means "Guaranteed non-throwing if and only if the expression in parentheses is true." `noexcept()` [used in a context where expressions are valid](https://en.cppreference.com/w/cpp/language/noexcept) is a boolean expression that's true if the expression in the parentheses would only involve `noexcept` calls, and false otherwise.

So this is the standard syntax for saying "this method is `noexcept` if and only if this expression is".

File: include/lsst/geom/CoordinateBase.h (hunk 5 above)
Comment (kfindeisen) [reply include/lsst/geom/CoordinateBase.h]: I did not have specific types in mind. However, we do allow any type as `T`, and if somebody did have a legitimate use for a `Point` of (presumably numeric) classes it would be very confusing if the program crashed because of an unexpected exception. Debugging exceptions thrown from `noexcept` methods is about as hard as debugging segfaults. :disappointed:

File: include/lsst/geom/CoordinateExpr.h (hunk 6 above)
Comment (r-owen) [reply include/lsst/geom/CoordinateExpr.h]: Please consider adding a brief code comment for a future reader, e.g. "the inner noexcept has a boolean value"

I realize it's a nuisance to document standard features

File: include/lsst/geom/LinearTransform.h [hunk 8]
```diff
@@ -156,34 +164,36 @@ class LinearTransform {
     /**
      * Return the determinant of the 2x2 matrix
      */
-    double computeDeterminant() const;
+    double computeDeterminant() const noexcept;
```
Comment (r-owen) [inline include/lsst/geom/LinearTransform.h]: Do you know that Eigen determinant() never throws?

File: include/lsst/geom/Extent.h (hunk 7 above)
Comment (r-owen) [reply include/lsst/geom/Extent.h]: I'm confused, again...I see operator== being marked `noexcept` for many of the classes in this package, just not this one.

File: include/lsst/geom/LinearTransform.h (hunk 8 above)
Comment (kfindeisen) [reply include/lsst/geom/LinearTransform.h]: Yes. As mentioned in the PR intro, `Eigen` has a policy of never throwing exceptions for anything except dynamic memory allocation.

File: include/lsst/geom/Extent.h (hunk 7 above)
Comment (kfindeisen) [reply include/lsst/geom/Extent.h]: I was hoping to address this after adding the `static_assert` on `N`...

This was another case of "this should always be computable, but the implementation is potentially throwing". The current implementation constructs a temporary `CoordinateExpr`; I think that requiring static Eigen matrices will make `CoordinateExpr` nothrow-constructible, sidestepping the problem.

**Comment by kfindeisen** [comment]
Hi @r-owen, can you take another look? Requiring `N > 0` in `CoordinateBase` simplifies the subclasses' exception specs enough that I could do most methods properly. In the process, I also removed all use of `noexcept(noexcept(...))`. :slightly_smiling_face:

File: include/lsst/geom/CoordinateBase.h [hunk 9]
```diff
@@ -47,15 +53,27 @@ class Extent;
 template <typename Derived, typename T, int N>
 class CoordinateBase {
 public:
+    static_assert(N > 0, "CoordinateBase must have a fixed positive length.");
```
Comment (r-owen) [inline include/lsst/geom/CoordinateBase.h]: Thank you for adding this. But...consider removing the word "fixed". I don't think it adds information and I worry it may be confusing.

File: include/lsst/geom/CoordinateBase.h [hunk 10]
```diff
@@ -110,12 +129,23 @@ class CoordinateBase<Derived, T, 2> {
     typedef T Element;
     static int const dimensions = 2;
     typedef Eigen::Matrix<T, 2, 1, Eigen::DontAlign> EigenVector;
-
```
Comment (r-owen) [inline include/lsst/geom/CoordinateBase.h]: Consider making the `allclose` methods `noexcept`. But if you'd rather not, that's fine. They are probably not used much, especially in production code.

File: include/lsst/geom/CoordinateExpr.h [hunk 11]
```diff
@@ -52,41 +52,41 @@ class CoordinateExpr : public CoordinateBase<CoordinateExpr<N>, bool, N> {
...
      *
      *  These operators do not provide interoperability with scalars.
      */
     //@{
-    CoordinateExpr and_(CoordinateExpr const& rhs) const;
-    CoordinateExpr or_(CoordinateExpr const& rhs) const;
-    CoordinateExpr not_() const;
+    CoordinateExpr and_(CoordinateExpr const& rhs) const noexcept;
+    CoordinateExpr or_(CoordinateExpr const& rhs) const noexcept;
+    CoordinateExpr not_() const noexcept;
```
Comment (r-owen) [inline include/lsst/geom/CoordinateExpr.h]: I would expect `expr[n]` to be unsafe (and thus not `noexcept`), since `n` may be out of bounds. Thus I would expect `and_`, `or_`, `not_` and the free functions `all` and `any` to be unsafe.

Certainly for `CoordinateBase` above you did not make the indexed accessors `noexcept`.

File: include/lsst/geom/Extent.h [hunk 12]
```diff
@@ -116,23 +128,27 @@ class ExtentBase : public CoordinateBase<Extent<T, N>, T, N> {
      *  No scalar interoperability is provided for Extent additive arithmetic operations.
      */
     //@{
-    Point<T, N> operator+(Point<T, N> const &other) const;
-    Extent<T, N> operator+(Extent<T, N> const &other) const {
+    Point<T, N> operator+(Point<T, N> const &other) const noexcept(Super::IS_ELEMENT_NOTHROW_COPYABLE);
+    Extent<T, N> operator+(Extent<T, N> const &other) const noexcept(Super::IS_ELEMENT_NOTHROW_COPYABLE) {
```
Comment (r-owen) [inline include/lsst/geom/Extent.h]: What if an integer value overflows or underflows for any of these math operations?

I see a few possibilities (if there is a problem):
- The math operations should not be made `noexcept`. That's simplest, of course.
- Only float versions should be made `noexcept`.

File: include/lsst/geom/Extent.h [hunk 13]
```diff
@@ -236,40 +258,44 @@ class Extent<T, 2> : public ExtentBase<T, 2> {
...
 
-    Extent(Extent const &other);
-    Extent(Extent &&other);
+    // Should be consistent with converting constructors
+    Extent(Extent const &other) = default;
+    Extent(Extent &&other) = default;
     ~Extent() = default;
 
     Extent &operator=(Extent const &other) = default;
     Extent &operator=(Extent &&other) = default;
```
Comment (r-owen) [inline include/lsst/geom/Extent.h]: Can these be `noexcept`? How about the destructor. Sorry if you already answered this elsewhere. 

(If you do make any changed, there's another instance of these below, and please look at `Point` as well).

File: include/lsst/geom/Extent.h [hunk 14]
```diff
@@ -285,59 +311,56 @@ class Extent<T, 3> : public ExtentBase<T, 3> {
...
     this->setX(static_cast<T>(other.getX()));
     this->setY(static_cast<T>(other.getY()));
 };
 
-// Should be consistent with converting constructor
-template <typename T>
-Extent<T, 2>::Extent(Extent<T, 2> const &) = default;
-template <typename T>
-Extent<T, 2>::Extent(Extent<T, 2> &&) = default;
-
```
Comment (r-owen) [inline include/lsst/geom/Extent.h]: Why did you remove these (and the 3-d equivalents)?

File: include/lsst/geom/SpherePoint.h [hunk 15]
```diff
@@ -65,7 +65,7 @@ class SpherePoint final {
      *                 interval [-&pi;/2, &pi;/2] radians.
      *
      * @throws pex::exceptions::InvalidParameterError
-     *         Thrown if `latitude` isout of range.
+     *         Thrown if `latitude` is out of range.
```
Comment (r-owen) [inline include/lsst/geom/SpherePoint.h]: Thank  you

File: include/lsst/geom/SpherePoint.h [hunk 16]
```diff
@@ -159,8 +159,10 @@ class SpherePoint final {
 
     /**
      * Make SpherePoint implicitly convertible to LonLat
+     *
+     * @exceptsafe Shall not throw exceptions.
```
Comment (r-owen) [inline include/lsst/geom/SpherePoint.h]: Do we need this if a method is marked "noexcept"? If the Doxygen show the "noexcept" in some useful way then  suggest removing these.

File: tests/test_box.py [hunk 17]
```diff
@@ -554,12 +557,17 @@ def _checkBoxConstruction(self, boxClass, size, center, precision):
             The maximum distance by which the box can be offset from ``center``.
         """
         msg = 'Box size = %s, location = %s' % (size, center)
-        box = boxClass.makeCenteredBox(center, size)
 
-        if all(size.gt(0)):
-            self._checkBoxProperties(box, size, center, precision, msg)
-        else:
-            self.assertTrue(box.isEmpty(), msg=msg)
+        if np.isfinite(center[0]) and np.isfinite(center[1]):
```
Comment (r-owen) [inline tests/test_box.py]: Consider `if all(np.isfinite(center)):`

File: tests/test_box.py [hunk 18]
```diff
@@ -554,12 +557,17 @@ def _checkBoxConstruction(self, boxClass, size, center, precision):
             The maximum distance by which the box can be offset from ``center``.
         """
         msg = 'Box size = %s, location = %s' % (size, center)
```
Comment (r-owen) [inline tests/test_box.py]: Pre-existing code, but consider possibly modernizing this code by replacing all use of `msg` in this code block using subTest: `with self.subTest(size=size, center=center):` followed by the rest of the code

File: tests/test_box.py [hunk 19]
```diff
@@ -554,12 +557,17 @@ def _checkBoxConstruction(self, boxClass, size, center, precision):
...
+        if np.isfinite(center[0]) and np.isfinite(center[1]):
+            box = boxClass.makeCenteredBox(center, size)
+            if all(size.gt(0)):
+                self._checkBoxProperties(box, size, center, precision, msg)
+            else:
+                self.assertTrue(box.isEmpty(), msg=msg)
+        elif boxClass == geom.Box2I:
+            with self.assertRaises(lsst.pex.exceptions.InvalidParameterError, msg=msg):
+                boxClass.makeCenteredBox(center, size)
+        # Don't want to throw for non-finite Box2D, but can't guarantee postconditions
```
Comment (r-owen) [inline tests/test_box.py]: I have no idea what this code comment means. Please clarify.

File: include/lsst/geom/CoordinateExpr.h (hunk 11 above)
Comment (kfindeisen) [reply include/lsst/geom/CoordinateExpr.h]: These methods are safe because they internally guarantee that only in-bounds indices will be used. Note that an attempt to `and_` two `CoordinateExpr` of different lengths will not compile.

File: include/lsst/geom/Extent.h (hunk 12 above)
Comment (kfindeisen) [reply include/lsst/geom/Extent.h]: Integer overflow (or, in fact, any operation on primitive types) does not normally raise exceptions.

I allowed for that possibility in `Box2I` because one of the `Box2I` constructors explicitly did so, so it would be reasonable to expect others to do so for consistency's sake. But I don't think it's likely that `Point`/`Extent` would ever have overflow checks added.

File: include/lsst/geom/Extent.h (hunk 13 above)
Comment (kfindeisen) [reply include/lsst/geom/Extent.h]: These are implicitly `noexcept` because they call only the base class constructor/assignment operator, which is `noexcept`. I could add a conditional `noexcept` specifier if you think messy* is better than implicit.

File: include/lsst/geom/Extent.h (hunk 14 above)
Comment (kfindeisen) [reply include/lsst/geom/Extent.h]: I moved them into the class definition. It turns out that `= default` only makes a constructor automatically `noexcept` if it's used when the constructor is first declared.

Of course, if I add explicit conditional `noexcept` everywhere, per your previous comment, then this argument becomes irrelevant.

File: include/lsst/geom/SpherePoint.h (hunk 16 above)
Comment (kfindeisen) [reply include/lsst/geom/SpherePoint.h]: It shows it, but IMO not well. The [method list](http://doxygen.lsst.codes/stack/doxygen/x_masterDoxyDoc/classlsst_1_1geom_1_1_sphere_point.html#pub-methods) has any method keywords verbatim, while the [detailed description](http://doxygen.lsst.codes/stack/doxygen/x_masterDoxyDoc/classlsst_1_1geom_1_1_sphere_point.html#a07e79418a7c071a99a590a5ae6f2ce13) has a small `noexcept` tag (which, it turns out, does not distinguish between conditional and unconditional `noexcept`).

I think this question came up when we were overhauling the C++ documentation standards...

File: tests/test_box.py (hunk 18 above)
Comment (kfindeisen) [reply tests/test_box.py]: @parejkoj told me that `subTest` doesn't work well with pytest.

File: tests/test_box.py (hunk 19 above)
Comment (kfindeisen) [reply tests/test_box.py]: This comment probably doesn't belong here; I'll remove it.

File: tests/test_box.py (hunk 18 above)
Comment (parejkoj) [reply tests/test_box.py]: Sadly, it does not: https://github.com/pytest-dev/pytest/issues/1367

Comment (r-owen) [reply tests/test_box.py]: Using subTest here will work the same with pytest as the current code. if the method in question is called by a loop then it will fail at the first failure either way. In this case subTest saves formatting a message and passing it around.

Comment (r-owen) [reply tests/test_box.py]: More generally, in my opinion writing unit tests that use subTest is perfectly acceptable even if pytest doesn't run all test cases. When we have pytest run code we want to know if it *passes*.  If it fails on any one iteration of a loop then, the build fails. The user can then run the test manually to see how many instances of the loop fails, if the user cares.

The reason I like subTest so much is that it can easily show me under what conditions a test fails, without the need to format messages and pass them around. I don't usually care if subTest stops the test on the first failure or shows all possible failures, since I have debugging ahead of me in either case. In fact if many iterations of a loop fails (a common situation if one fails) then running all of them can be a bit of a nuisances because it may make it difficult to estimate how many unique bugs are present.

Comment (parejkoj) [reply tests/test_box.py]: Ah, I hadn't considered the message-management aspect of using `subTest`.

I'm assuming the `with subTest` block goes in the calling function, not this one?

Comment (r-owen) [reply tests/test_box.py]: I would probably put `subTest` here if multiple other bits of code call this function. But it could go in the callers instead. I think it will work identically either way.

Comment (r-owen) [reply tests/test_box.py]: Try running the test with python instead of pytest. You should see the information appended to the FAIL message. In my opinion that's good enough -- tests normally don't fail and when they do we can use python to get additional info. Presumably pytest will eventually be updated to show the info as well.

Comment (kfindeisen) [reply tests/test_box.py]: Yes, they work when run as `python [test file]`, but I don't agree that it's reasonable to constantly switch between pytest and unittest like that. I'd rather stick with the old approach than force anyone debugging the code to adopt a more convoluted and less obvious workflow.

Comment (parejkoj) [reply tests/test_box.py]: I agree with @kfindeisen here: we don't know when pytest will be updated (if ever: that issue is over 2 years old now), and a user isn't going to know to switch to running python to get the useful error messages.
